### PR TITLE
Add ability to group forms into wizards.

### LINKED ISSUE #17: Ability to group forms into wizards (Session Wizard)
Ability to group forms into wizards

### PR DESCRIPTION
**Pull request by phalamag**
This will resolve #17

### CODE REVIEW

File: formfactory/actions.py [hunk 1]
```diff
@@ -98,3 +98,9 @@ def login(form_instance, **kwargs):
     )
     if user is not None:
         auth.login(request, user)
+
+
+@register
+def wizard_done_step(form_dict, **kwargs):
```
Comment (qoda) [inline formfactory/actions.py]: Explain the real world use of this please.

File: formfactory/admin.py [hunk 2]
```diff
@@ -23,11 +23,13 @@ class FormFieldInline(admin.StackedInline):
     form = forms.FormFieldAdminForm
     model = models.FormField
     prepopulated_fields = {"slug": ["title"]}
+    extra = 1
```
Comment (qoda) [inline formfactory/admin.py]: Please remove, leave as default (3)

File: formfactory/admin.py [hunk 3]
```diff
@@ -23,11 +23,13 @@ class FormFieldInline(admin.StackedInline):
...
 
 
 class FormActionThroughInline(admin.StackedInline):
     form = forms.FormActionThroughAdminForm
     model = models.FormActionThrough
+    extra = 1
```
Comment (qoda) [inline formfactory/admin.py]: Please remove, leave as default (3)

File: formfactory/admin.py [hunk 4]
```diff
@@ -50,7 +52,28 @@ class FormDataAdmin(admin.ModelAdmin):
     readonly_fields = utils.get_all_model_fields(models.FormData)
 
 
+class FormThroughInline(admin.StackedInline):
+    form = forms.FormThroughAdminForm
+    model = models.FormOrderThrough
+    extra = 1
```
Comment (qoda) [inline formfactory/admin.py]: Please remove, leave as default (3)

File: formfactory/admin.py [hunk 5]
```diff
@@ -50,7 +52,28 @@ class FormDataAdmin(admin.ModelAdmin):
...
+
+
+class WizardActionThroughInline(admin.StackedInline):
+    form = forms.WizardActionThroughAdminForm
+    model = models.WizardActionThrough
+    extra = 1
```
Comment (qoda) [inline formfactory/admin.py]: Please remove, leave as default (3)

File: formfactory/forms.py [hunk 6]
```diff
@@ -59,3 +59,25 @@ class FormActionParamAdminForm(forms.ModelForm):
     class Meta:
         model = models.ActionParam
         fields = ["key", "value"]
+
+
+class WizardAdminForm(forms.ModelForm):
+    class Meta:
+        model = models.Wizard
+        fields = ["title", "slug", ]
```
Comment (qoda) [inline formfactory/forms.py]: Redundant trailing comma in list, this is only typical of a one item tuple, i.e `("item 1", )`

Comment (qoda) [inline formfactory/forms.py]: Also are those the only fields that can be modified?

File: formfactory/models.py [hunk 7]
```diff
@@ -183,3 +184,60 @@ class Meta:
...
+    title = models.CharField(
+        max_length=30, help_text=_("A short descriptive title.")
+    )
+    slug = models.SlugField(
+        max_length=40, db_index=True, unique=True
+    )
+    forms = models.ManyToManyField(Form, through="FormOrderThrough")
+    actions = models.ManyToManyField(Action, through="WizardActionThrough")
+
+    @cached_property
```
Comment (qoda) [inline formfactory/models.py]: If this property is not absolutely necessary rather don't query in a property. This means that an additional query will be made every time this object is instantiated. Also note cached_properties only cache for the length an object is instantiated while may be a very short time.

File: formfactory/models.py [hunk 8]
```diff
@@ -183,3 +184,60 @@ class Meta:
...
+    def form_list(self):
+        return self.forms.all()
+
+    @cached_property
```
Comment (qoda) [inline formfactory/models.py]: Same as above

File: formfactory/urls.py [hunk 9]
```diff
@@ -1,8 +1,13 @@
...
 
-from formfactory.views import FactoryFormView
+from formfactory.views import FactoryFormView, FactoryWizardView
 
 
 urlpatterns = [
-    url(r"^(?P<slug>[-\w]+)/$", FactoryFormView.as_view(), name="form-detail")
+    url(r"^(?P<slug>[-\w]+)/$", FactoryFormView.as_view(), name="form-detail"),
+    url(
+        r"^profile/(?P<slug>[\w-]+)(?:/(?P<step>[\w-]+))?/$",
```
Comment (qoda) [inline formfactory/urls.py]: Why profile?

File: formfactory/models.py [hunk 10]
```diff
@@ -183,3 +184,60 @@ class Meta:
...
+    def as_wizard(self):
+        return {
+            "form_list": self.form_list,
+            "actions": self.actions.all(),
+        }
+
+
+class FormOrderThrough(models.Model):
```
Comment (qoda) [inline formfactory/models.py]: How are these through models exposed in admin?

File: formfactory/views.py [hunk 11]
```diff
@@ -45,3 +49,82 @@ def get_success_url(self):
...
+class FactoryWizardView(NamedUrlSessionWizardView):
+    form_list = [EmptyForm, ]
+    redirect_to = None
+
+    def get_prefix(self, request, *args, **kwargs):
+        return "%s-%s" % (self.__class__.__name__, kwargs["slug"])
+
+    def dispatch(self, request, *args, **kwargs):
+        wizard_slug = kwargs.get("slug")
+        import pdb;pdb.set_trace()
```
Comment (qoda) [inline formfactory/views.py]: Remove!

File: formfactory/forms.py (hunk 6 above)
Comment (phalamag) [inline formfactory/forms.py]: At this point, yes

File: formfactory/models.py (hunk 10 above)
Comment (phalamag) [inline formfactory/models.py]: The through models are `inlines` on the `wizard` model admin.

File: formfactory/urls.py (hunk 9 above)
Comment (phalamag) [inline formfactory/urls.py]: Erroneous copy-and-paste

File: formfactory/actions.py (hunk 1 above)
Comment (phalamag) [inline formfactory/actions.py]: I was using this to experiment with what should happen at the end of a wizard. Removing it now.

**Pull request merged**